### PR TITLE
Add a method to set/unset multiple config values with one write

### PR DESCRIPTION
**Pull request by nickvergessen**
Related #12785

Based on the idea from https://github.com/owncloud/core/pull/13556#issuecomment-71110473

Currently when we set "n" configs in one go, we write the file "n"-times.
Instead we should reduce it almost "1". E.g.: when changing the SMTP settings now only 1 write action is performed instead of 8-10.

With less write actions, we might break the config.php less often. Worth a try.

@LukasReschke @MorrisJobke @DeepDiver1975 requested second version

### CODE REVIEW

**Comment by nickvergessen** [comment]
Also fixed a potentially undefined var in lib/private/setup/oci.php, causing the retab, so use whitespace free diff there:
https://github.com/owncloud/core/pull/13621/files?w=1#diff-8

**Comment by nickvergessen** [comment]
@owncloud-bot re-test this please

File: core/command/db/converttype.php [hunk 1]
```diff
@@ -289,10 +289,12 @@ protected function saveDBInfo(InputInterface $input) {
...
-		$this->config->setSystemValue('dbname', $dbname);
-		$this->config->setSystemValue('dbhost', $dbhost);
-		$this->config->setSystemValue('dbuser', $username);
-		$this->config->setSystemValue('dbpassword', $password);
+		$this->config->setSystemValues([
+			'dbtype'		=> $type,
+			'dbname'		=> $dbname,
+			'dbhost'		=> $dbhost,
+			'dbuser'		=> $username,
+			'dbpassword'	=> $password,
```
Comment (LukasReschke) [inline core/command/db/converttype.php]: Is there any deeper sense behind the different indentation above?

Comment (nickvergessen) [inline core/command/db/converttype.php]: github displayes 8space tabs while the code uses 4space tabs. It's all fine in the editor.

Comment (LukasReschke) [inline core/command/db/converttype.php]: Indeed: https://github.com/owncloud/core/pull/13621/files?ts=4 – so true.

File: lib/private/legacy/config.php [hunk 2]
```diff
@@ -59,6 +59,16 @@ public static function setValue($key, $value) {
 	}
 
 	/**
+	 * Sets and deletes values and writes the config.php
+	 *
+	 * @param array $configs Associative array with `key => value` pairs
+	 *                       If value is null, the config key will be deleted
+	 */
+	public static function setValues(array $configs) {
```
Comment (DeepDiver1975) [inline lib/private/legacy/config.php]: I'd not add this and change related code segments to use \OC::$server->getConfig()

Comment (nickvergessen) [inline lib/private/legacy/config.php]: The problem is, that this is only used in the database install parts, and I don't want to break it with a circular dependency :see_no_evil:

**Comment by MorrisJobke** [comment]
@owncloud-bot re-test this please

File: lib/private/legacy/config.php (hunk 2 above)
Comment (DeepDiver1975) [inline lib/private/legacy/config.php]: agreed - then don't touch that part at all - :speak_no_evil:

**Comment by DeepDiver1975** [comment]
failing unit tests:
````
 OC\Settings\Controller\MailSettingsControllerTest.testSetMailSettings
 Stack Trace

OC\Settings\Controller\MailSettingsControllerTest::testSetMailSettings
Expectation failed for method name is equal to <string:setSystemValue> when invoked 15 time(s).
Method was expected to be called 15 times, actually called 0 times.

	0.183941	1
 OC\Settings\Controller\MailSettingsControllerTest.testStoreCredentials
 Stack Trace

OC\Settings\Controller\MailSettingsControllerTest::testStoreCredentials
Expectation failed for method name is equal to <string:setSystemValue> when invoked 2 time(s).
Method was expected to be called 2 times, actually called 0 times.

File: lib/private/legacy/config.php (hunk 2 above)
Comment (nickvergessen) [inline lib/private/legacy/config.php]: I can still add a deprecated message or something like that,
or we even throw an exception if this is called and catch it in the db stuff

File: tests/settings/controller/mailsettingscontrollertest.php [hunk 3]
```diff
@@ -69,26 +69,37 @@ public function testSetMailSettings() {
...
+		$config->expects($this->exactly(2))
+			->method('setSystemValues');
 		/**
 		 * FIXME: Use the following block once Jenkins uses PHPUnit >= 4.1
-		 */
-		/*
-		$this->container['Config']
-			->expects($this->exactly(3))
-			->method('deleteSystemValue')
 			->withConsecutive(
```
Comment (nickvergessen) [inline tests/settings/controller/mailsettingscontrollertest.php]: nice github displayes unknown syntax that is in comments :-O

**Comment by nickvergessen** [comment]
tests should pass now

File: lib/private/legacy/config.php (hunk 2 above)
Comment (DeepDiver1975) [inline lib/private/legacy/config.php]: on setup the risk for parallel writes is not that critical

**Comment by LukasReschke** [comment]
Nice stuff :+1:

**Comment by DeepDiver1975** [comment]
Unit tests on oracle succeed:

```sh
deepdiver@alien:~/Development/ownCloud/core-autotest [system-config-multiset]$ ./autotest.sh oci
Using database oc_autotest
Setup environment for oci testing ...
drop the database

Benutzer wurde gelöscht.

create the database

Benutzer wurde erstellt.


Benutzer wurde geändert.


Benutzerzugriff (Grant) wurde erteilt.

INDEX
END INDEX
Testing with oci ...
PHPUnit 4.3.5 by Sebastian Bergmann.

Configuration read from /home/deepdiver/Development/ownCloud/core-autotest/tests/phpunit-autotest.xml

..............S..............................................   61 / 2841 (  2%)
.............................................................  122 / 2841 (  4%)
......................F......................................  183 / 2841 (  6%)
.............................................................  244 / 2841 (  8%)
.............................................................  305 / 2841 ( 10%)
.............................................................  366 / 2841 ( 12%)
.............................................................  427 / 2841 ( 15%)
.............................................................  488 / 2841 ( 17%)
.............................................................  549 / 2841 ( 19%)
..............................S.SSSSSSS......................  610 / 2841 ( 21%)
.............................................................  671 / 2841 ( 23%)
.............................................................  732 / 2841 ( 25%)
.............................................................  793 / 2841 ( 27%)
.....SSSSSS..................................................  854 / 2841 ( 30%)
......................................SSSSSSSSSSSSSSSSSSSSSSS  915 / 2841 ( 32%)
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS..........................  976 / 2841 ( 34%)
............................................................. 1037 / 2841 ( 36%)
............................................................. 1098 / 2841 ( 38%)
............................................................. 1159 / 2841 ( 40%)
............................................................. 1220 / 2841 ( 42%)
............................................................. 1281 / 2841 ( 45%)
............................................................. 1342 / 2841 ( 47%)
............................................................. 1403 / 2841 ( 49%)
............................................................. 1464 / 2841 ( 51%)
............................................................. 1525 / 2841 ( 53%)
............................................................. 1586 / 2841 ( 55%)
............................................................. 1647 / 2841 ( 57%)
............................................................. 1708 / 2841 ( 60%)
............................................................. 1769 / 2841 ( 62%)
............................................................. 1830 / 2841 ( 64%)
.....................................SS...........SSSSSSSSSSS 1891 / 2841 ( 66%)
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS...................... 1952 / 2841 ( 68%)
.......SS.................................................... 2013 / 2841 ( 70%)
............................................................. 2074 / 2841 ( 73%)
............................................................. 2135 / 2841 ( 75%)
............................................................. 2196 / 2841 ( 77%)
............................................................. 2257 / 2841 ( 79%)
............................................................. 2318 / 2841 ( 81%)
............................................................. 2379 / 2841 ( 83%)
...................S..................SS.......S............. 2440 / 2841 ( 85%)
............................................................. 2501 / 2841 ( 88%)
............................................................. 2562 / 2841 ( 90%)
............................................................. 2623 / 2841 ( 92%)
............................................................. 2684 / 2841 ( 94%)
............................................................. 2745 / 2841 ( 96%)
............................................................. 2806 / 2841 ( 98%)
...................................

Time: 22.75 minutes, Memory: 214.00Mb

There was 1 failure:

1) Test_Appconfig::testGetApps
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
-    0 => '123456'
-    1 => 'anotherapp'
-    2 => 'core'
-    3 => 'files'
-    4 => 'files_encryption'
-    5 => 'files_sharing'
-    6 => 'files_trashbin'
-    7 => 'files_versions'
-    8 => 'provisioning_api'
-    9 => 'someapp'
-    10 => 'testapp'
-    11 => 'user_ldap'
+    0 => 'anotherapp'
+    1 => 'core'
+    2 => 'files'
+    3 => 'files_encryption'
+    4 => 'files_sharing'
+    5 => 'files_trashbin'
+    6 => 'files_versions'
+    7 => 'provisioning_api'
+    8 => 'someapp'
+    9 => 'testapp'
+    10 => 'user_ldap'
+    11 => '123456'
 )

/home/deepdiver/Development/ownCloud/core-autotest/tests/lib/appconfig.php:51

--

There were 131 skipped tests:

**Comment by DeepDiver1975** [comment]
:+1:

File: lib/private/setup/abstractdatabase.php [hunk 4]
```diff
@@ -35,20 +35,24 @@ public function validate($config) {
...
+		]);
+
+		$this->dbuser = $dbUser;
+		$this->dbpassword = $dbPass;
+		$this->dbname = $dbName;
+		$this->dbhost = $dbHost;
+		$this->tableprefix = $dbTablePrefix;
 	}
+
+	abstract public function setupDatabase($userName);
```
Comment (LukasReschke) [inline lib/private/setup/abstractdatabase.php]: PHPDoc? :) – Scrutinizer unhappy :(

**Pull request merged**